organic_scoring/organic_scoring_base.py

```diff
@@ -0,0 +1,316 @@
+
+        Args:
+            synapse: The synapse to handle.
+
+        Returns:
+            bt.StreamingSynapse: The handled synapse.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _query_miners(self, sample: Any) -> dict[str, Any]:
+        """Query the miners with a sample
+
+        Args:
+            sample: The sample to query with.
+
+        Returns:
+            dict[str, Any]: The responses from the miners.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _generate_rewards(self, sample: Any, responses: Sequence[Any], reference: Any = None) -> dict[str, Any]:
+        """Generate rewards based on the sample and responses.
+
+        Args:
+            sample: The sample to use.
+            responses: The responses from the miners.
+            reference: The reference data. Defaults to None.
+
+        Returns:
+            dict[str, Any]: The generated rewards information.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def _set_weights(self, rewards: dict[str, Any]):
+        """Set the weights for the miners.
+
+        Args:
+            rewards: Dict with rewards and any additional info.
+        """
+        raise NotImplementedError
+
+    async def _generate_reference(self, sample: Any) -> Optional[Any]:
+        """Generate a reference based on the sample.
+
+        Args:
+            sample: The sample used to generate the reference.
+
+        Returns:
+            Optional[Any]: The generated reference, if any.
+        """
+        return None
+
+    async def _log_results(
+        self,
+        logs: dict[str, Any],
+        reference: Any,
+        responses: dict[str, Any],
+        rewards: dict[str, Any],
+        sample: Any,
+        *args,
+        **kwargs,
+    ) -> dict[str, Any]:
+        """Log the results of the organic scoring iteration.
+
+        Args:
+            logs: The logs to record. Default values in the dict:
+                - "organic_time_sample": Time taken in seconds to sample the organic queue or synthetic dataset;
+                - "organic_time_responses": Time taken in seconds to query the miners and generate reference;
+                - "organic_time_rewards": Time taken in seconds to generate rewards;
+                - "organic_time_weights": Time taken in seconds to set the weights;
+                - "organic_time_total": Total time taken in seconds for the iteration;
+                - "organic_queue_len": Current length of the organic queue;
+                - "is_organic_sample": If the sample is from the organic queue.
+            reference: The reference data.
+            responses: The responses from the miners.
+            rewards: The generated rewards.
+            sample: The sample used.
+
+        Returns:
+            dict[str, Any]: The logs recorded.
+        """
+        return logs
+
+    async def _priority_fn(self, synapse: bt.Synapse) -> float:
```

Comment:
Same for the following functions, they raise notimplementederror but are not marked as abstract